### DISCUSSION TITLE
Multicompiler and HMR

### DISCUSSION THREAD

**Discussion by fivethreeo** [webpack/webpack#14361] [Q&A]
In razzle we use two webpack configs that runs two compilers in dev one for client and one for node. First the client compiler runs and creates a assets.json, then the server compiler runs and starts  watching for changes, then we start webpack dev server for the client.

This sort of works, but what would be better would be to use multicompiler in webpack dev server.  But the issue is then: How can we get the client to compile first to generate the assets.json. Then compile the server. Then HMR reload the server. Then HMR reload the client.

Been looking into the webpack source. I think I can make the server compiler wait. But how can I make the HMR happen in the right order?

**Comment by alexander-akait** [comment]
https://webpack.js.org/configuration/other-options/#dependencies

Allow to build server `client`, then `server`

> If you want to manually recompile the bundle, navigating to the /invalidate route will invalidate the current compilation of the bundle and recompile it for you via webpack-dev-middleware. Depending on your configuration, the URL may look like http://localhost:9000/invalidate.

https://webpack.js.org/configuration/dev-server/

For reloading client in HMR, anyway you can create simple plugin and use `compiler.invalidate` then you need this

**Reply by fivethreeo** [reply]
Ah, great. But the HMR part I don't quite get. Invalidate makes the HMR reload? How can I invalidate the other compiler? Do I need to disable anything so it does not happen automatically? Will it work with react-refresh?  

Do I understand this correctly or are these silly questions? 😉

**Reply by alexander-akait** [reply]
> Invalidate makes the HMR reload?

Yes, if hash of build was changed, it will apply new modules or reload the page, even more you can rerun build using `invalidate`, dev server just react on that

> How can I invalidate the other compiler?

Small example https://github.com/webpack/webpack-dev-middleware/blob/master/src/index.js#L66

> Do I need to disable anything so it does not happen automatically?

Can you clarify?

**Comment by fivethreeo** [comment]
I wonder if this is already supported in webpack-dev-middleware. I just need a way for webpack-dev-middleware to wait for the server(s)  to start before doing any reloading. I can send a signal from the server(s). But how can I make the middleware wait for those signal(s). I looked into the source and see it does a Promise that resolves with ready if serverSideRender is true. What would be great was if I could supply my own Promise.

**Reply by fivethreeo** [reply]
Hm, looking at what I wrote, ready resolves after compilation and HMR and next() only runs the next handler. So that made no sense. But just no HMR until servers signal or servers can be reached.
